v3_async_wip/v3_async_wip/iothub_session.py

```diff
@@ -245,9 +271,37 @@ async def desired_property_updates(
         """Returns an async generator of incoming twin desired property patches"""
         await self._mqtt_client.enable_twin_patch_receive()
         try:
-            yield self._mqtt_client.incoming_twin_patches
+            yield self._add_disconnect_interrupt_to_generator(
+                self._mqtt_client.incoming_twin_patches
+            )
         finally:
-            await self._mqtt_client.disable_twin_patch_receive()
+            try:
+                await self._mqtt_client.disable_twin_patch_receive()
+            except mqtt.MQTTError:
+                # i.e. not connected
+                pass
+
+    def _add_disconnect_interrupt_to_generator(
+        self, generator: AsyncGenerator[_T, None]
+    ) -> AsyncGenerator[_T, None]:
+        """Wrap a generator in another generator that will either return the next item yielded by
+        the original generator, or raise error in the event of disconnect
+        """
+
+        async def wrapping_generator():
+            while True:
+                new_item_t = asyncio.create_task(generator.__anext__())
+                done, _ = await asyncio.wait(
+                    [new_item_t, self._report_conn_drop_task],
```

Comment:
>_report_conn_drop_task [](http://example.com/codeflow?start=38&length=22)

Does this work if you're cancelling 2 generators? I know that you can't await an already-awaited task, but I don't know how this applies to asyncio.wait.

Reply:
If this is a problem, you could use asyncio.Event here (e.g. `done, pending = asyncio.wait([new_item_t,, self.drop_conn_event.wait()], return_when=asyncio.FIRST_COMPLETED)`. 

The hidden trick here is that you have to cancel the `drop_con_event,wait()` task if it's inside the `pending` array.  If you don't do this, you'll leak a task every time this gets called.

Reply:
I think it does, but I'll look into it and double check. I had considered doing this (not with an Event, but with a Future), but my implementation seemed simpler.

Reply:
Update: This does work fine for cancelling multiple generators